### DISCUSSION TITLE
TCP handshake testcase

### DISCUSSION THREAD

**Discussion by crossedfinger-j** [ANLAB-KAIST/KENSv3#191] [Project Q&A 2. Transport Layer (TCP)]
in tcp handshake test case, 
<img width="641" alt="스크린샷 2024-04-09 오전 11 33 25" src="https://github.com/ANLAB-KAIST/KENSv3/assets/103436153/f603f8ee-c6a6-4696-9a2d-182a38c2d4ff">
line number = 665 -> is that the case CONNECT in my tcpAssignment.cpp?
In my case, cannot print any log in case CONNECT when I run the ./test-kens-handshake.

**Comment by Pusnow** [comment]
It depends on test cases.

For test cases staring with `TestAccept_`, your server-side codes (`accept()`, etc.) are tested.
For test cases staring with `TestConnect_`, your client-side codes (`connect()`, etc.) are tested.

Run `./test-kens-handshake --gtest_list_tests` to list all test cases.

**Reply by crossedfinger-j** [reply]
thanks for your answer. and I think the connect function when I test TestAccept_, return -1 (as the printf log).
I think the connect function is not my syscall CONNECT. 
but I have ack packet and store it in backlogList.
Is it possible situation? (I use timer when 3-way handshakng).

**Reply by Pusnow** [reply]
> the connect function when I test TestAccept_, return -1 (as the printf log).

It fails if the connection establishment is failed. If you implement server-side system calls correctly, it will success.

> the connect function is not my syscall CONNECT.

`TestAceept_` uses TA's implementation for clients, so it will not call your `connect()` implementation.

> but I have ack packet and store it in backlogList.
Is it possible situation? (I use timer when 3-way handshakng).

If you receive SYN and response it by sending SYN/ACK, the client will send ACK.

**Comment by crossedfinger-j** [comment]
Thank you for your answer.

I mean, I checked ACK receivied from the client. 

But, connect() function in test code returns -1.

I think there is some reason connect() function returns error, but I couldn't know the reason since it is implemented by TAs.

If the client sends ACK for the last step of three-way handshake, doesn't this indicate successful connect() ?

If not, can you give me some cases connect() function returns -1?

Thank you for your kind answer!

**Reply by crossedfinger-j** [reply]
I also checked the packets for SYN (1st step of handshake) along with ACK (the 3rd step of handshake)  from the client in function packetArrived, which made any wrong (all good).

**Reply by Pusnow** [reply]
Are all the `connect()`s failing? Which test cases are you running?

Some test cases intentionally make `connect()`s to be failed.

**Reply by crossedfinger-j** [reply]
in testhandshake.cpp, 

 test ret : -1 
/Users/jinlee/Desktop/kens/KENSv3/app/kens/testhandshake.cpp:167: Failure
Expected equality of these values:
  (int)client_sockets.size()
    Which is: 0
  expected_connect
    Which is: 1
    <img width="685" alt="스크린샷 2024-04-09 오후 8 57 32" src="https://github.com/ANLAB-KAIST/KENSv3/assets/103436153/d0facfac-fed9-4e22-934b-a4c8dbb69157">

sorry for the late reply :(

**Reply by Pusnow** [reply]
Would you send your source code to us? cs341.pj@kaist.ac.kr

**Reply by crossedfinger-j** [reply]
I send the email! Thanks

**Reply by Pusnow** [reply]
`uint32_t sendAckNum = htonl(ntohl(seqNum + 1));` -> `uint32_t sendAckNum = htonl(seqNum + 1);`

1. You have already called `ntohl` and store the result into `seqNum`.
2. Also, you should add 1 to host ordered integer not network ordered integer.

You can use pcap files to check that you've set right SEQ/ACK numbers.

```bash
tshark  -o tcp.relative_sequence_numbers:FALSE  -r  TestAccept_Backlog1.pcap 
```
